File: src/Tests/UnitTests/serialization_unit_tests.cpp
//-- includes -----
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#include "BinarySerializer.h"
#include "BinaryDeserializer.h"
#include "BinaryUtility.h"
#include "JsonSerializer.h"
#include "JsonDeserializer.h"
#include "MathUtility.h"
#include "SerializableList.h"
#include "SerializableMap.h"

#include "serialization_unit_tests.h"
#include "serialization_unit_tests.rfks.h"
#include "unit_test.h"

//-- public interface -----
bool run_serialization_unit_tests()
{
	UNIT_TEST_MODULE_BEGIN("serialization")
		UNIT_TEST_MODULE_CALL_TEST(serialization_utility_test_endian_swap);
		UNIT_TEST_MODULE_CALL_TEST(serialization_utility_test_reflection_from_json);
		UNIT_TEST_MODULE_CALL_TEST(serialization_utility_test_reflection_from_bytes);
	UNIT_TEST_MODULE_END()
}

//-- private functions -----
void test_endian_mode(Serialization::Endian endianMode)
{
	uint8_t scratch[32];

	int16_t expectedValue16 = 0x1234;
	Serialization::write_int16((uint8_t*)&scratch, expectedValue16, endianMode);
	int16_t actualValue16 = Serialization::read_int16((uint8_t*)&scratch, endianMode);
	assert(actualValue16 == expectedValue16);

	int32_t expectedValue24 = 0x123456;
	Serialization::write_int24((uint8_t*)&scratch, expectedValue24, endianMode);
	int32_t actualValue24 = Serialization::read_int24((uint8_t*)&scratch, endianMode);
	assert(actualValue24 == expectedValue24);

	int32_t expectedValue32 = 0x12345678;
	Serialization::write_int32((uint8_t*)&scratch, expectedValue32, endianMode);
	int32_t actualValue32 = Serialization::read_int32((uint8_t*)&scratch, endianMode);
	assert(actualValue32 == expectedValue32);

	int64_t expectedValue64 = 0x1234567890abcdef;
	Serialization::write_int64((uint8_t*)&scratch, expectedValue64, endianMode);
	int64_t actualValue64 = Serialization::read_int64((uint8_t*)&scratch, endianMode);
	assert(actualValue64 == expectedValue64);

	float expectedFloat = k_real_pi;
	Serialization::write_float((uint8_t*)&scratch, expectedFloat, endianMode);
	float actualFloat = Serialization::read_float((uint8_t*)&scratch, endianMode);
	assert(is_nearly_equal(expectedFloat, actualFloat, k_real_epsilon));

	double expectedDouble = k_real64_pi;
	Serialization::write_double((uint8_t*)&scratch, expectedDouble, endianMode);
	double actualDouble = Serialization::read_double((uint8_t*)&scratch, endianMode);
	assert(is_double_nearly_equal(expectedDouble, actualDouble, k_real64_epsilon));
}

bool serialization_utility_test_endian_swap()
{
	UNIT_TEST_BEGIN("endian swap")

		test_endian_mode(Serialization::Endian::Little);
		test_endian_mode(Serialization::Endian::Big);

	UNIT_TEST_COMPLETE()
}

void build_serialization_test_struct(SerializationTestStruct& outStruct)
{

	Serialization::BoolList boolArray;
	boolArray.push_back(true);
	boolArray.push_back(false);
	boolArray.push_back(true);

	Serialization::List<int> intArray({{1, 2, 3}});

	Serialization::List<SerializationPoint2dStruct> pointArray;
	pointArray.push_back({1.2345f, 5.4321f});
	pointArray.push_back({5.4321f, 1.2345f});

	Serialization::Map<int, SerializationPoint2dStruct> intPointMap;
	intPointMap.insert({1, {1.2345f, 5.4321f}});
	intPointMap.insert({2, {5.4321f, 1.2345f}});

	Serialization::Map<std::string, SerializationPoint2dStruct> stringPointMap;
	stringPointMap.insert({"key1", {1.2345f, 5.4321f}});
	stringPointMap.insert({"key2", {5.4321f, 1.2345f}});

	auto point3d_ptr = std::make_shared<SerializationPoint3dStruct>();
	point3d_ptr->x_field = 1.2345f;
	point3d_ptr->y_field = 5.4321f;
	point3d_ptr->z_field = 9.8765f;

	outStruct.bool_field= true;
	outStruct.byte_field= -123;
	outStruct.ubyte_field= 123;
	outStruct.short_field= -1234;
	outStruct.ushort_field= 1234;
	outStruct.int_field= -123456;
	outStruct.uint_field= 123456;
	outStruct.long_field= -123456789;
	outStruct.float_field= 1.2345f;
	outStruct.double_field= 1.23456789;
	outStruct.string_field= Serialization::String("hello");
	outStruct.enum_field= SerializationTestEnum_Value2;
	outStruct.point2d_field= {1.2345f, 5.4321f};
	outStruct.point_ptr_field= Serialization::ObjectPtr<SerializationPointStruct>(point3d_ptr);
	outStruct.bool_array= boolArray;
	outStruct.int_array= intArray;
	outStruct.point2d_array= pointArray;
	outStruct.int_point_map= intPointMap;
	outStruct.string_point_map= stringPointMap;
}

void verify_serialization_test_struct(const SerializationTestStruct& actual, const SerializationTestStruct& expected)
{
	assert(actual.bool_field == expected.bool_field);
	assert(actual.byte_field == expected.byte_field);
	assert(actual.ubyte_field == expected.ubyte_field);
	assert(actual.short_field == expected.short_field);
	assert(actual.ushort_field == expected.ushort_field);
	assert(actual.int_field == expected.int_field);
	assert(actual.uint_field == expected.uint_field);
	assert(actual.long_field == expected.long_field);
	assert(is_nearly_equal(actual.float_field, expected.float_field, k_real_epsilon));
	assert(is_double_nearly_equal(actual.double_field, expected.double_field, k_real64_epsilon));
	assert(actual.string_field == expected.string_field);
	assert(actual.enum_field == expected.enum_field);
	assert(is_nearly_equal(actual.point2d_field.x_field, expected.point2d_field.x_field, k_real_epsilon));
	assert(is_nearly_equal(actual.point2d_field.y_field, expected.point2d_field.y_field, k_real_epsilon));

	auto expected_point3d= 
		std::dynamic_pointer_cast<SerializationPoint3dStruct>(
			expected.point_ptr_field.getSharedPointer());
	auto actual_point3d= 
		std::dynamic_pointer_cast<SerializationPoint3dStruct>(
			actual.point_ptr_field.getSharedPointer());
	assert(expected_point3d);
	assert(actual_point3d);
	assert(is_nearly_equal(expected_point3d->x_field, actual_point3d->x_field, k_real_epsilon));
	assert(is_nearly_equal(expected_point3d->y_field, actual_point3d->y_field, k_real_epsilon));
	assert(is_nearly_equal(expected_point3d->z_field, actual_point3d->z_field, k_real_epsilon));

	assert(actual.bool_array.size() == expected.bool_array.size());
	for (size_t i = 0; i < actual.bool_array.size(); ++i)
	{
		assert(actual.bool_array[i] == expected.bool_array[i]);
	}

	assert(actual.int_array.size() == expected.int_array.size());
	for (size_t i = 0; i < actual.int_array.size(); ++i)
	{
		assert(actual.int_array[i] == expected.int_array[i]);
	}

	assert(actual.point2d_array.size() == expected.point2d_array.size());
	for (size_t i = 0; i < 2; ++i)
	{
		const auto& actualPoint = actual.point2d_array[i];
		const auto& expectedPoint = expected.point2d_array[i];

		assert(is_nearly_equal(actualPoint.x_field, expectedPoint.x_field, k_real_epsilon));
		assert(is_nearly_equal(actualPoint.y_field, expectedPoint.y_field, k_real_epsilon));
	}

	assert(actual.int_point_map.size() == expected.int_point_map.size());
	for (const auto& pair : actual.int_point_map)
	{
		const int key = pair.first;
		const auto& actualPoint = pair.second;
		const auto iter = expected.int_point_map.find(key);
		const auto& expectedPoint = iter->second;

		assert(is_nearly_equal(actualPoint.x_field, expectedPoint.x_field, k_real_epsilon));
		assert(is_nearly_equal(actualPoint.y_field, expectedPoint.y_field, k_real_epsilon));
	}

	assert(actual.string_point_map.size() == expected.string_point_map.size());
	for (const auto& pair : actual.string_point_map)
	{
		const std::string& key = pair.first;
		const auto& actualPoint = pair.second;
		const auto iter = expected.string_point_map.find(key);
		const auto& expectedPoint = iter->second;

		assert(is_nearly_equal(actualPoint.x_field, expectedPoint.x_field, k_real_epsilon));
		assert(is_nearly_equal(actualPoint.y_field, expectedPoint.y_field, k_real_epsilon));
	}
}

bool serialization_utility_test_reflection_from_json()
{
	UNIT_TEST_BEGIN("reflect from json")
		SerializationTestStruct expected = {};
		build_serialization_test_struct(expected);

		std::string jsonString;
		bool bCanSerialize= Serialization::serializeToJsonString(expected, jsonString);
		assert(bCanSerialize);

		SerializationTestStruct actual= {};
		bool bCanDeserialize = Serialization::deserializeFromJsonString(jsonString, actual);

		verify_serialization_test_struct(actual, expected);

	UNIT_TEST_COMPLETE()
}

bool serialization_utility_test_reflection_from_bytes()
{
	UNIT_TEST_BEGIN("reflect from bytes")
		SerializationTestStruct expected;
		build_serialization_test_struct(expected);

		std::vector<uint8_t> bytes;
		bool bCanSerialize= Serialization::serializeToBytes(expected, bytes);
		assert(bCanSerialize);

		SerializationTestStruct actual= {};
		bool bCanDeserialize = Serialization::deserializeFromBytes(bytes, actual);

		verify_serialization_test_struct(actual, expected);
	UNIT_TEST_COMPLETE()
}

File: bindings/csharp/Tests/SerializationUnitTests.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace MikanXR
{
	enum SerializationTestEnum
	{
		Value1,
		Value2,
		Value3,
	};

	class SerializationPoint : PolymorphicStruct
	{
		public static readonly long classId= 0;
	};

	class SerializationPoint2d : SerializationPoint
	{
		public static new readonly long classId= 1;

		public SerializationPoint2d()
		{
		}

		public SerializationPoint2d(float x, float y)
		{
			x_field = x;
			y_field = y;
		}

		public float x_field = 0;
		public float y_field = 0;
	};

	class SerializationPoint3d : SerializationPoint
	{
		public static new readonly long classId= 2;

		public SerializationPoint3d()
		{
		}

		public SerializationPoint3d(float x, float y, float z)
		{
			x_field = x;
			y_field = y;
			z_field = z;
		}

		public float x_field = 0;
		public float y_field = 0;
		public float z_field = 0;
	};

	class SerializationTestObject
	{
		public bool bool_field;
		public sbyte byte_field;
		public byte ubyte_field;
		public short short_field;
		public ushort ushort_field;
		public int int_field;
		public uint uint_field;
		public long long_field;
		public float float_field;
		public double double_field;
		public string string_field;
		public SerializationTestEnum enum_field;
		public SerializationPoint2d point2d_field;
		public PolymorphicObject point_ptr_field;
		public PolymorphicObject null_ptr_field;
		public List<bool> bool_array;
		public List<int> int_array;
		public List<float> float_array;
		public List<string> string_array;
		public List<SerializationPoint2d> point2d_array;
		public Dictionary<int, SerializationPoint2d> int_point_map;
		public Dictionary<string, SerializationPoint2d> string_point_map;
	};

	public class SerializationUnitTests
	{
		SerializationTestObject buildSerializationTestObject()
		{
			SerializationTestObject testObject = new SerializationTestObject();
			var boolArray = new List<bool>() { true, false, true };
			var intArray = new List<int>() {1, 2, 3};
			var floatArray = new List<float>() {1.2345f, 5.4321f, 9.8765f};
			var stringArray = new List<string>() {"hello", "world", "!"};

			var pointArray = new List<SerializationPoint2d>();
			pointArray.Add( new SerializationPoint2d(1.2345f, 5.4321f) );
			pointArray.Add( new SerializationPoint2d(5.4321f, 1.2345f) );

			var intPointMap = new Dictionary<int, SerializationPoint2d>();
			intPointMap.Add(1, new SerializationPoint2d(1.2345f, 5.4321f));
			intPointMap.Add(2, new SerializationPoint2d(5.4321f, 1.2345f));

			var stringPointMap = new Dictionary<string, SerializationPoint2d>();
			stringPointMap.Add("key1", new SerializationPoint2d(1.2345f, 5.4321f));
			stringPointMap.Add("key2", new SerializationPoint2d(5.4321f, 1.2345f));

			testObject.bool_field= true;
			testObject.byte_field= -123;
			testObject.ubyte_field= 123;
			testObject.short_field= -1234;
			testObject.ushort_field= 1234;
			testObject.int_field= -123456;
			testObject.uint_field= 123456;
			testObject.long_field= -123456789;
			testObject.float_field= 1.2345f;
			testObject.double_field= 1.23456789;
			testObject.string_field= "hello";
			testObject.enum_field= SerializationTestEnum.Value2;
			testObject.point2d_field= new SerializationPoint2d(1.2345f, 5.4321f);
			testObject.point_ptr_field = new PolymorphicObject();
			testObject.point_ptr_field.setInstance(new SerializationPoint3d(1.2345f, 5.4321f, 9.8765f));
			testObject.null_ptr_field = new PolymorphicObject();
			testObject.bool_array= boolArray;
			testObject.int_array= intArray;
			testObject.float_array = floatArray;
			testObject.string_array = stringArray;
			testObject.point2d_array= pointArray;
			testObject.int_point_map= intPointMap;
			testObject.string_point_map= stringPointMap;

			return testObject;
		}

		bool verifySerializationTestObject(
			SerializationTestObject actual,
			SerializationTestObject expected)
		{
			bool success;

			success = (actual.bool_field == expected.bool_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.byte_field == expected.byte_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.ubyte_field == expected.ubyte_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.short_field == expected.short_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.ushort_field == expected.ushort_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.int_field == expected.int_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.uint_field == expected.uint_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.long_field == expected.long_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (Math.Abs(actual.float_field - expected.float_field) <= float.Epsilon);
			Debug.Assert(success);
			if (!success) return false;

			success = (Math.Abs(actual.double_field - expected.double_field) <= double.Epsilon);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.string_field == expected.string_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.enum_field == expected.enum_field);
			Debug.Assert(success);
			if (!success) return false;

			success = (Math.Abs(actual.point2d_field.x_field - expected.point2d_field.x_field) <= float.Epsilon);
			Debug.Assert(success);
			if (!success) return false;

			success = (Math.Abs(actual.point2d_field.y_field - expected.point2d_field.y_field) <= float.Epsilon);
			Debug.Assert(success);
			if (!success) return false;

			var expected_point3d= (SerializationPoint3d)expected.point_ptr_field.Instance;
			var actual_point3d= (SerializationPoint3d)actual.point_ptr_field.Instance;

			success = (Math.Abs(expected_point3d.x_field - actual_point3d.x_field) <= float.Epsilon);
			Debug.Assert(success);
			if (!success) return false;

			success = (Math.Abs(expected_point3d.y_field - actual_point3d.y_field) <= float.Epsilon);
			Debug.Assert(success);
			if (!success) return false;

			success = (Math.Abs(expected_point3d.z_field - actual_point3d.z_field) <= float.Epsilon);
			Debug.Assert(success);
			if (!success) return false;

			success = (expected.null_ptr_field.Instance == null);
			Debug.Assert(success);
			if (!success) return false;

			success = (expected.null_ptr_field.RuntimeClassId == 0);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.null_ptr_field.Instance == null);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.null_ptr_field.RuntimeClassId == 0);
			Debug.Assert(success);
			if (!success) return false;

			success = (actual.bool_array.Count == expected.bool_array.Count);
			Debug.Assert(success);
			if (!success) return false;

			for (int i = 0; i < actual.bool_array.Count; ++i)
			{
				success = (actual.bool_array[i] == expected.bool_array[i]);
				Debug.Assert(success);
				if (!success) return false;
			}

			success = (actual.int_array.Count == expected.int_array.Count);
			Debug.Assert(success);
			if (!success) return false;

			for (int i = 0; i < actual.int_array.Count; ++i)
			{
				success = (actual.int_array[i] == expected.int_array[i]);
				Debug.Assert(success);
				if (!success) return false;
			}

			success = (actual.float_array.Count == expected.float_array.Count);
			Debug.Assert(success);
			if (!success) return false;

			for (int i = 0; i < actual.float_array.Count; ++i)
			{
				success = (Math.Abs(actual.float_array[i] - expected.float_array[i]) <= float.Epsilon);
				Debug.Assert(success);
				if (!success) return false;
			}

			success = (actual.string_array.Count == expected.string_array.Count);
			Debug.Assert(success);
			if (!success) return false;

			for (int i = 0; i < actual.string_array.Count; ++i)
			{
				success = (actual.string_array[i] == expected.string_array[i]);
				Debug.Assert(success);
				if (!success) return false;
			}

			success = (actual.point2d_array.Count == expected.point2d_array.Count);
			Debug.Assert(success);
			if (!success) return false;

			for (int i = 0; i < 2; ++i)
			{
				var actualPoint = actual.point2d_array[i];
				var expectedPoint = expected.point2d_array[i];

				success = (Math.Abs(actualPoint.x_field - expectedPoint.x_field) <= float.Epsilon);
				Debug.Assert(success);
				if (!success) return false;

				success = (Math.Abs(actualPoint.y_field - expectedPoint.y_field) <= float.Epsilon);
				Debug.Assert(success);
				if (!success) return false;
			}

			success = (actual.int_point_map.Count == expected.int_point_map.Count);
			Debug.Assert(success);
			if (!success) return false;

			foreach (var pair in actual.int_point_map)
			{
				var key = pair.Key;
				var actualPoint = pair.Value;
				var expectedPoint = expected.int_point_map[key];

				success = (Math.Abs(actualPoint.x_field - expectedPoint.x_field) <= float.Epsilon);
				Debug.Assert(success);
				if (!success) return false;

				success = (Math.Abs(actualPoint.y_field - expectedPoint.y_field) <= float.Epsilon);
				Debug.Assert(success);
				if (!success) return false;
			}

			success = (actual.string_point_map.Count == expected.string_point_map.Count);
			Debug.Assert(success);
			if (!success) return false;

			foreach (var pair in actual.string_point_map)
			{
				var key = pair.Key;
				var actualPoint = pair.Value;
				var expectedPoint = expected.string_point_map[key];

				success = (Math.Abs(actualPoint.x_field - expectedPoint.x_field) <= float.Epsilon);
				Debug.Assert(success);
				if (!success) return false;

				success = (Math.Abs(actualPoint.y_field - expectedPoint.y_field) <= float.Epsilon);
				Debug.Assert(success);
				if (!success) return false;
			}

			return true;
		}

		public bool TestReflectionFromJson()
		{
			bool success = true;

			try
			{
				var expected = buildSerializationTestObject();

				string jsonString = JsonSerializer.serializeToJsonString(expected);
				var actual = new SerializationTestObject();
				bool bCanDeserialize= JsonDeserializer.deserializeFromJsonString(jsonString, actual);

				if (!bCanDeserialize)
				{
					success = false;
				}
				else
				{
					success = verifySerializationTestObject(actual, expected);
				}
			}
			catch (Exception e)
			{
				Console.WriteLine("    Exception: " + e.Message);
				success = false;
			}

			Console.WriteLine("    TestReflectionFromJson - " + (success ? "PASSED" : "FAILED"));
			return success;
		}

		public bool TestReflectionFromBytes()
		{
			bool success = true;

			try
			{
				var expected = buildSerializationTestObject();

				byte[] bytes = BinarySerializer.SerializeToBytes(expected);

				var actual = new SerializationTestObject();
				bool bCanDeserialize = BinaryDeserializer.DeserializeFromBytes(bytes, actual);

				if (!bCanDeserialize)
				{
					success = false;
				}
				else
				{
					success = verifySerializationTestObject(actual, expected);
				}
			}
			catch (Exception e)
			{
				Console.WriteLine("    Exception: " + e.Message);
				success = false;
			}

			Console.WriteLine("    TestReflectionFromBytes - " + (success ? "PASSED" : "FAILED"));
			return success;
		}

		public bool RunAllTests()
		{
			bool success = true;
			Console.WriteLine("[SerializationUnitTests]");

			success &= TestReflectionFromJson();
			success &= TestReflectionFromBytes();

			Console.WriteLine("  SerializationUnitTests module - " + (success ? "PASSED" : "FAILED"));
			return success;
		}
	}
}